lib/pbench/test/unit/server/database/test_users_db.py

```diff
@@ -0,0 +1,152 @@
+        """
+        Fixture to mock a DB session for testing.
+
+        We patch the SQLAlchemy db_session to our fake session. We also store a
+        server configuration object directly on the Database.Base (normally
+        done during DB initialization) because that can't be monkeypatched.
+        """
+        __class__.session = FakeSession(User)
+        monkeypatch.setattr(Database, "db_session", __class__.session)
+        Database.Base.config = server_config
+        yield monkeypatch
+
+    @staticmethod
+    def add_dummy_user(fake_db):
+        dummy_user = User(
+            id=str(uuid.uuid4()),
+            username="dummy",
+        )
+        dummy_user.add()
+        return dummy_user
+
+    def test_construct(self, fake_db):
+        """Test User db contructor"""
+        user = self.add_dummy_user(fake_db)
+        assert user.username == "dummy"
+
+        expected_commits = [FakeRow.clone(user)]
+        self.session.check_session(queries=0, committed=expected_commits)
+        self.session.reset_context()
+
+    def test_is_admin(self, fake_db):
+        uuid4 = str(uuid.uuid4())
+        user = User(id=uuid4, username="dummy_admin", roles=["ADMIN"])
+        user.add()
+        assert user.is_admin()
+        assert user.id == uuid4
+        user1 = User(id="1", username="non_admin")
+        user1.add()
+        assert not user1.is_admin()
+
+        expected_commits = [FakeRow.clone(user), FakeRow.clone(user1)]
+        self.session.check_session(queries=0, committed=expected_commits)
+        self.session.reset_context()
+
+    def test_user_survives_dataset_real_session(self, db_session, create_user):
+        """The User isn't automatically removed when the referenced
+        dataset is deleted.
+        """
+        user = create_user
+        ds = Dataset(owner=user.username, name="fio", resource_id="deadbeef")
+        ds.add()
+        ds.delete()
+        with pytest.raises(DatasetNotFound):
+            Dataset.query(resource_id=ds.resource_id)
+        assert User.query(username=user.username)
+
+    def test_user_survives_dataset_delete(self, fake_db):
+        """The User isn't automatically removed when the referenced
+        dataset is deleted.
+        """
+        self.add_dummy_user(fake_db)
+        user = User.query(username="dummy")
+
+        expected_commits = [FakeRow.clone(user)]
+        self.session.check_session(
+            queries=1, committed=expected_commits, filters=["username=dummy"]
+        )
+
+        ds = Dataset(owner=user.username, name="fio", resource_id="deadbeef")
+        ds.add()
+        ds_row = FakeRow.clone(ds)
+        expected_commits.append(ds_row)
+        self.session.check_session(
+            queries=1, committed=expected_commits, filters=["username=dummy"]
+        )
+
+        ds.delete()
+        expected_commits.remove(ds_row)
+        self.session.check_session(queries=0, committed=expected_commits)
+        assert user
```

Comment:
Again, you're only asserting that the proxy object (which we haven't touched) still exists. (It will.) That's not a very strong test! You need to check whether the row is still in the DB, which means a query.

(Although I'm not entirely sure this test case is meaningful since my simplistic faked DB engine doesn't handle cascade deletes across tables anyway!)

Reply:
Yeah, that's true. It doesn't accomplish anything in terms of what we are trying to test.